src/Ads/AccountService.php

```diff
@@ -215,46 +216,52 @@ public function get_billing_status(): array {
 	/**
 	 * Gets the ads account access status.
 	 *
-	 * @return bool
+	 * @return array
 	 * @throws Exception When email is not present or invite link is empty when email present and has no account access.
 	 */
 	public function get_ads_account_has_access() {
-		try {
-			// Check if ads id is present.
-			$ads_id = $this->options->get_ads_id();
-
-			if ( empty( $ads_id ) ) {
-				return [
-					'has_access'  => false,
-					'invite_link' => '',
-				];
-			}
+		// Check if ads id is present.
+		$ads_id = $this->options->get_ads_id();
 
-			$connection_status  = $this->container->get( Connection::class )->get_status();
-			$email              = $connection_status['email'] ?? '';
-			$accept_invite_link = $this->options->get( OptionsInterface::ADS_BILLING_URL, '' );
+		if ( empty( $ads_id ) ) {
+			throw new Exception( __( 'Ads id not present', 'google-listings-and-ads' ) );
+		}
 
-			// If no email, means google account is not connected.
-			if ( empty( $email ) ) {
-				throw new Exception( 'Google account is not connected' );
-			}
+		$connection_status  = $this->container->get( Connection::class )->get_status();
+		$email              = $connection_status['email'] ?? '';
+		$accept_invite_link = $this->options->get( OptionsInterface::ADS_BILLING_URL, '' );
 
-			$status = $this->container->get( Ads::class )->has_access( $email );
-			$status = boolval( $status );
+		// If no email, means google account is not connected.
+		if ( empty( $email ) ) {
+			throw new Exception( __( 'Google account is not connected', 'google-listings-and-ads' ) );
+		}
 
-			// If no access to ads account and there is no invite_link, throw exception.
-			if ( ! $status && empty( $accept_invite_link ) ) {
-				throw new Exception( 'Billing URL is not present' );
-			}
+		$status = $this->container->get( Ads::class )->has_access( $email );
+
+		// If no access to ads account and there is no invite_link, throw exception.
+		if ( ! $status && empty( $accept_invite_link ) ) {
+			throw new Exception( __( 'Billing URL is not present', 'google-listings-and-ads' ) );
+		}
+
+		// If we have access, complete the step so that it won't be called next time.
+		if ( $status ) {
+			$this->state->complete_step( 'account_access' );
```

Comment:
Won't this be taken care of during foreach loop of the `setup_account()` method above as long as we aren't throwing an exception?

Reply:
This is partially correct. `get_ads_account_has_access` function is being called from two places.

1. `/ads/accounts` endpoint, which calls `setup_account` method and mark each step as `complete` after success. 
2. `/ads/account-status` endpoint which checks whether we have access to ads account.

Second one will be called during polling and while it is called, if we have access then we should mark `account_access` step as done because next time `setup_account` method is called, it won't call the Ads API for access status unnecessarily; saving us an extra request and time.